src/lcm/dispatchers.py

```diff
@@ -99,19 +110,24 @@ def productmap(func, variables):
             might be additional dimensions.
 
     """
-    func = allow_args(func)  # vmap cannot deal with keyword-only arguments
+    func = allow_args(func)  # jax.vmap cannot deal with keyword-only arguments
 
-    if len(variables) != len(set(variables)):
-        raise ValueError("Same argument provided more than once.")
+    if duplicates := {v for v in variables if variables.count(v) > 1}:
+        raise ValueError(
+            f"Same argument provided more than once in variables: {duplicates}",
+        )
 
     signature = inspect.signature(func)
-    vmapped = _product_map(func, variables)
-    vmapped.__signature__ = signature
+    vmapped = _base_productmap(func, variables)
+
+    # This raises a mypy error but is perfectly fine to do. See
+    # https://github.com/python/mypy/issues/12472
+    vmapped.__signature__ = signature  # type: ignore[attr-defined]
 
     return allow_kwargs(vmapped)
 
 
-def vmap_1d(func, variables):
+def vmap_1d(func: F, variables: list[str], *, apply_allow_kwargs: bool = True) -> F:
     """Apply vmap such that func is mapped over the specified variables.
 
     In contrast to vmap, vmap_1d preserves the function signature and allows the
```

Comment:
Yes. It is less flexible than `jax.vmap` in terms of vectorization because it is designed to only be called when we have sparse variables where we call the dispatched function on array inputs, and we do not want a Cartesian product but a row-wise execution.